### DISCUSSION TITLE
pydantic.AfterValidator is ignored for path parameter types

### DISCUSSION THREAD

**Discussion by sjdv1982** [fastapi/fastapi#10131] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
##########################################################
# Pydantic section (working)
# (adapted from the Pydantic documentation)
##########################################################

from typing_extensions import Annotated

from pydantic.functional_validators import AfterValidator
from pydantic import TypeAdapter

def check_squares(v: int) -> int:
    assert v**0.5 % 1 == 0, f'{v} is not a square number'
    return v

MyNumber = Annotated[int,  AfterValidator(check_squares)]

ta = TypeAdapter(MyNumber)
ta.validate_python(9)    # OK
# ta.validate_python(8)  # gives a validation error, as it should

##########################################################
# FastAPI section (not working as expected)
##########################################################

from fastapi import FastAPI
app = FastAPI()

@app.get("/{number}")
async def get_number(number: MyNumber):
    # The next line should be redundant: if we reach this point,
    # Pydantic validation ought to have been performed already
    try:
        TypeAdapter(MyNumber).validate_python(number)
    except Exception:
        return -1
    return number
```

### Description

* Open the browser and call the endpoint '/8'
* I would expect an error message, as '8' is not valid input
* However, '-1' is returned, showing that the `get_number` coroutine is being entered
* Therefore,  `pydantic.AfterValidator` is not being triggered for the MyNumber type

### Operating System

Linux

### FastAPI Version

0.101.1

### Pydantic Version

2.2.1

### Python Version

3.11.4

**Comment by JavierSanchezCastro** [comment, marked as answer]
Hi!

You forgot to declare MyNumber as Path parameter

```python3
##########################################################
# Pydantic section (working)
# (adapted from the Pydantic documentation)
##########################################################

from typing_extensions import Annotated

from pydantic.functional_validators import AfterValidator
from pydantic import TypeAdapter

def check_squares(v: int) -> int:
    assert v**0.5 % 1 == 0, f'{v} is not a square number'
    return v

MyNumber = Annotated[int,  AfterValidator(check_squares)]

ta = TypeAdapter(MyNumber)
ta.validate_python(9)    # OK
# ta.validate_python(8)  # gives a validation error, as it should

##########################################################
# FastAPI section (not working as expected)
##########################################################

from fastapi import FastAPI, Path
app = FastAPI()

MyNumber = Annotated[int, AfterValidator(check_squares), Path()]

@app.get("/{number}")
async def get_number(number: MyNumber):
    return number
```
or

```python3
##########################################################
# Pydantic section (working)
# (adapted from the Pydantic documentation)
##########################################################

from typing_extensions import Annotated

from pydantic.functional_validators import AfterValidator
from pydantic import TypeAdapter

def check_squares(v: int) -> int:
    assert v**0.5 % 1 == 0, f'{v} is not a square number'
    return v

MyNumber = Annotated[int,  AfterValidator(check_squares)]

ta = TypeAdapter(MyNumber)
ta.validate_python(9)    # OK
# ta.validate_python(8)  # gives a validation error, as it should

##########################################################
# FastAPI section (not working as expected)
##########################################################

from fastapi import FastAPI, Path
app = FastAPI()

@app.get("/{number}")
async def get_number(number: Annotated[MyNumber, Path()]):
    return number
```

**Reply by sjdv1982** [reply]
That works, thank you. 
I am a bit puzzled why it is necessary, given `app.get({number}`, but it works.

**Comment by wallneradam** [comment]
I have this problem as well. I think it should work without defining `Path()`. If you just add an `int` or `str`, it can find out if it is a path or query parameter. It can just annotate it automatically.

I could create a pydantic field class with `__get_pydantic_core_schema__` and `__get_pydantic_json_schema__`, and it just works without annotating it with`Path` or `Query`, but the new pydantic Annotated way would be much more readable and nice, if it could work with FastAPI out of the box.
